{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# UNC HiSeq mRNAseq gene expression (RSEM)\n",
    "\n",
    "The goal of this notebook is to introduce you to the mRNAseq gene expression BigQuery table.\n",
    "\n",
    "This table contains all available TCGA Level-3 gene expression data produced by UNC's RNAseqV2 pipeline using the Illumina HiSeq platform, as of July 2016.  The most recent archive (*eg* ``unc.edu_BRCA.IlluminaHiSeq_RNASeqV2.Level_3.1.11.0``) for each of the 33 tumor types was downloaded from the DCC, and data extracted from all files matching the pattern ``%.rsem.genes.normalized_results``. Each of these raw “RSEM genes normalized results” files has two columns: gene_id and normalized_count.  The gene_id string contains two parts: the gene symbol, and the Entrez gene ID, separated by **|**  *eg*: **`TP53|7157`**.  During ETL, the gene_id string is split and the gene symbol is stored in the ``original_gene_symbol`` field, and the Entrez gene ID is stored in the ``gene_id`` field.  In addition, the Entrez ID is used to look up the current HGNC approved gene symbol, which is stored in the ``HGNC_gene_sybmol`` field. \n",
    "\n",
    "In order to work with BigQuery, you need to import the python bigquery module (`gcp.bigquery`) and you need to know the name(s) of the table(s) you are going to be working with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import gcp.bigquery as bq\n",
    "mRNAseq_BQtable = bq.Table('isb-cgc:tcga_201607_beta.mRNA_UNC_HiSeq_RSEM')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From now on, we will refer to this table using this variable ($mRNAseq_BQtable), but we could just as well explicitly give the table name each time.\n",
    "\n",
    "Let's start by taking a look at the table schema:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bqsv\" id=\"1_146983051636\"></div>\n",
       "    <script>\n",
       "      require(['extensions/bigquery', 'element!1_146983051636', 'style!/static/extensions/bigquery.css'],\n",
       "        function(bq, dom) {\n",
       "          bq.renderSchema(dom, [{\"type\": \"STRING\", \"name\": \"ParticipantBarcode\"}, {\"type\": \"STRING\", \"name\": \"SampleBarcode\"}, {\"type\": \"STRING\", \"name\": \"AliquotBarcode\"}, {\"type\": \"STRING\", \"name\": \"Study\"}, {\"type\": \"STRING\", \"name\": \"SampleTypeLetterCode\"}, {\"type\": \"STRING\", \"name\": \"Platform\"}, {\"type\": \"STRING\", \"name\": \"original_gene_symbol\"}, {\"type\": \"STRING\", \"name\": \"HGNC_gene_symbol\"}, {\"type\": \"INTEGER\", \"name\": \"gene_id\"}, {\"type\": \"FLOAT\", \"name\": \"normalized_count\"}]);\n",
       "        }\n",
       "      );\n",
       "    </script>\n",
       "    "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%bigquery schema --table $mRNAseq_BQtable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's count up the number of unique patients, samples and aliquots mentioned in this table.  We will do this by defining a very simple parameterized query.  (Note that when using a variable for the table name in the FROM clause, you should not also use the square brackets that you usually would if you were specifying the table name as a string.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%sql --module count_unique\n",
    "\n",
    "DEFINE QUERY q1\n",
    "SELECT COUNT (DISTINCT $f, 25000) AS n\n",
    "FROM $t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " There are   9530 unique values in the field ParticipantBarcode. \n",
      " There are  10289 unique values in the field SampleBarcode. \n",
      " There are  10291 unique values in the field AliquotBarcode. \n"
     ]
    }
   ],
   "source": [
    "fieldList = ['ParticipantBarcode', 'SampleBarcode', 'AliquotBarcode']\n",
    "for aField in fieldList:\n",
    "  field = mRNAseq_BQtable.schema[aField]\n",
    "  rdf = bq.Query(count_unique.q1,t=mRNAseq_BQtable,f=field).results().to_dataframe()\n",
    "  print \" There are %6d unique values in the field %s. \" % ( rdf.iloc[0]['n'], aField)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can do the same thing to look at how many unique gene symbols and gene ids exist in the table:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " There are  20501 unique values in the field original_gene_symbol. \n",
      " There are  20182 unique values in the field HGNC_gene_symbol. \n",
      " There are  20531 unique values in the field gene_id. \n"
     ]
    }
   ],
   "source": [
    "fieldList = ['original_gene_symbol', 'HGNC_gene_symbol', 'gene_id']\n",
    "for aField in fieldList:\n",
    "  field = mRNAseq_BQtable.schema[aField]\n",
    "  rdf = bq.Query(count_unique.q1,t=mRNAseq_BQtable,f=field).results().to_dataframe()\n",
    "  print \" There are %6d unique values in the field %s. \" % ( rdf.iloc[0]['n'], aField)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on the counts, we can see that there are a few instances where the original gene symbol (from the underlying TCGA data file), or the HGNC gene symbol or the gene id (also from the original TCGA data file) is missing, but for the majority of genes, all three values should be available and for the most part the original gene symbol and the HGNC gene symbol that was added during ETL should all match up.  This next query will generate the complete list of genes for which none of the identifiers are null, and where the original gene symbol and the HGNC gene symbol match.  This list has over 18000 genes in it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bqtv\" id=\"2_146983056142\"><table><tr><th>HGNC_gene_symbol</th><th>original_gene_symbol</th><th>gene_id</th></tr><tr><td>A1BG</td><td>A1BG</td><td>1</td></tr><tr><td>A1CF</td><td>A1CF</td><td>29974</td></tr><tr><td>A2M</td><td>A2M</td><td>2</td></tr><tr><td>A2ML1</td><td>A2ML1</td><td>144568</td></tr><tr><td>A4GALT</td><td>A4GALT</td><td>53947</td></tr><tr><td>A4GNT</td><td>A4GNT</td><td>51146</td></tr><tr><td>AAAS</td><td>AAAS</td><td>8086</td></tr><tr><td>AACS</td><td>AACS</td><td>65985</td></tr><tr><td>AADAC</td><td>AADAC</td><td>13</td></tr><tr><td>AADACL2</td><td>AADACL2</td><td>344752</td></tr><tr><td>AADACL3</td><td>AADACL3</td><td>126767</td></tr><tr><td>AADACL4</td><td>AADACL4</td><td>343066</td></tr><tr><td>AADAT</td><td>AADAT</td><td>51166</td></tr><tr><td>AAGAB</td><td>AAGAB</td><td>79719</td></tr><tr><td>AAK1</td><td>AAK1</td><td>22848</td></tr><tr><td>AAMP</td><td>AAMP</td><td>14</td></tr><tr><td>AANAT</td><td>AANAT</td><td>15</td></tr><tr><td>AARS</td><td>AARS</td><td>16</td></tr><tr><td>AARS2</td><td>AARS2</td><td>57505</td></tr><tr><td>AARSD1</td><td>AARSD1</td><td>80755</td></tr><tr><td>AASDH</td><td>AASDH</td><td>132949</td></tr><tr><td>AASDHPPT</td><td>AASDHPPT</td><td>60496</td></tr><tr><td>AASS</td><td>AASS</td><td>10157</td></tr><tr><td>AATF</td><td>AATF</td><td>26574</td></tr><tr><td>AATK</td><td>AATK</td><td>9625</td></tr></table></div>\n",
       "    <br />(rows: 18153, time: 25.5s,     4GB processed, job: job_q9uLYjnsTKOT4eFAyWM1uu2VVPw)<br />\n",
       "    <script>\n",
       "      require(['extensions/charting', 'element!2_146983056142', 'style!/static/extensions/charting.css'],\n",
       "        function(charts, dom) {\n",
       "          charts.render(dom,\n",
       "            {\n",
       "              chartStyle:\"paged_table\",\n",
       "              dataName:\"0\",\n",
       "              fields:\"HGNC_gene_symbol,original_gene_symbol,gene_id\",\n",
       "              totalRows:18153,\n",
       "              rowsPerPage:25,\n",
       "            }, {}, {\"rows\": [{\"c\": [{\"v\": \"A1BG\"}, {\"v\": \"A1BG\"}, {\"v\": 1}]}, {\"c\": [{\"v\": \"A1CF\"}, {\"v\": \"A1CF\"}, {\"v\": 29974}]}, {\"c\": [{\"v\": \"A2M\"}, {\"v\": \"A2M\"}, {\"v\": 2}]}, {\"c\": [{\"v\": \"A2ML1\"}, {\"v\": \"A2ML1\"}, {\"v\": 144568}]}, {\"c\": [{\"v\": \"A4GALT\"}, {\"v\": \"A4GALT\"}, {\"v\": 53947}]}, {\"c\": [{\"v\": \"A4GNT\"}, {\"v\": \"A4GNT\"}, {\"v\": 51146}]}, {\"c\": [{\"v\": \"AAAS\"}, {\"v\": \"AAAS\"}, {\"v\": 8086}]}, {\"c\": [{\"v\": \"AACS\"}, {\"v\": \"AACS\"}, {\"v\": 65985}]}, {\"c\": [{\"v\": \"AADAC\"}, {\"v\": \"AADAC\"}, {\"v\": 13}]}, {\"c\": [{\"v\": \"AADACL2\"}, {\"v\": \"AADACL2\"}, {\"v\": 344752}]}, {\"c\": [{\"v\": \"AADACL3\"}, {\"v\": \"AADACL3\"}, {\"v\": 126767}]}, {\"c\": [{\"v\": \"AADACL4\"}, {\"v\": \"AADACL4\"}, {\"v\": 343066}]}, {\"c\": [{\"v\": \"AADAT\"}, {\"v\": \"AADAT\"}, {\"v\": 51166}]}, {\"c\": [{\"v\": \"AAGAB\"}, {\"v\": \"AAGAB\"}, {\"v\": 79719}]}, {\"c\": [{\"v\": \"AAK1\"}, {\"v\": \"AAK1\"}, {\"v\": 22848}]}, {\"c\": [{\"v\": \"AAMP\"}, {\"v\": \"AAMP\"}, {\"v\": 14}]}, {\"c\": [{\"v\": \"AANAT\"}, {\"v\": \"AANAT\"}, {\"v\": 15}]}, {\"c\": [{\"v\": \"AARS\"}, {\"v\": \"AARS\"}, {\"v\": 16}]}, {\"c\": [{\"v\": \"AARS2\"}, {\"v\": \"AARS2\"}, {\"v\": 57505}]}, {\"c\": [{\"v\": \"AARSD1\"}, {\"v\": \"AARSD1\"}, {\"v\": 80755}]}, {\"c\": [{\"v\": \"AASDH\"}, {\"v\": \"AASDH\"}, {\"v\": 132949}]}, {\"c\": [{\"v\": \"AASDHPPT\"}, {\"v\": \"AASDHPPT\"}, {\"v\": 60496}]}, {\"c\": [{\"v\": \"AASS\"}, {\"v\": \"AASS\"}, {\"v\": 10157}]}, {\"c\": [{\"v\": \"AATF\"}, {\"v\": \"AATF\"}, {\"v\": 26574}]}, {\"c\": [{\"v\": \"AATK\"}, {\"v\": \"AATK\"}, {\"v\": 9625}]}], \"cols\": [{\"type\": \"string\", \"id\": \"HGNC_gene_symbol\", \"label\": \"HGNC_gene_symbol\"}, {\"type\": \"string\", \"id\": \"original_gene_symbol\", \"label\": \"original_gene_symbol\"}, {\"type\": \"number\", \"id\": \"gene_id\", \"label\": \"gene_id\"}]});\n",
       "        }\n",
       "      );\n",
       "    </script>\n",
       "  "
      ],
      "text/plain": [
       "QueryResultsTable job_q9uLYjnsTKOT4eFAyWM1uu2VVPw"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "SELECT\n",
    "  HGNC_gene_symbol,\n",
    "  original_gene_symbol,\n",
    "  gene_id\n",
    "FROM\n",
    "  $mRNAseq_BQtable\n",
    "WHERE\n",
    "  ( original_gene_symbol IS NOT NULL\n",
    "    AND HGNC_gene_symbol IS NOT NULL\n",
    "    AND original_gene_symbol=HGNC_gene_symbol\n",
    "    AND gene_id IS NOT NULL )\n",
    "GROUP BY\n",
    "  original_gene_symbol,\n",
    "  HGNC_gene_symbol,\n",
    "  gene_id\n",
    "ORDER BY\n",
    "  HGNC_gene_symbol"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We might also want to know how often the gene symbols do not agree:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bqtv\" id=\"3_146983056775\"><table><tr><th>HGNC_gene_symbol</th><th>original_gene_symbol</th><th>gene_id</th></tr><tr><td>A1BG-AS1</td><td>NCRNA00181</td><td>503538</td></tr><tr><td>A2M-AS1</td><td>LOC144571</td><td>144571</td></tr><tr><td>AACSP1</td><td>AACSL</td><td>729522</td></tr><tr><td>AADACP1</td><td>LOC201651</td><td>201651</td></tr><tr><td>AAED1</td><td>C9orf21</td><td>195827</td></tr><tr><td>AAMDC</td><td>C11orf67</td><td>28971</td></tr><tr><td>AAR2</td><td>C20orf4</td><td>25980</td></tr><tr><td>AARD</td><td>C8orf85</td><td>441376</td></tr><tr><td>AATBC</td><td>LOC284837</td><td>284837</td></tr><tr><td>AATK-AS1</td><td>LOC388428</td><td>388428</td></tr><tr><td>ABHD11-AS1</td><td>WBSCR26</td><td>171022</td></tr><tr><td>ABHD16A</td><td>BAT5</td><td>7920</td></tr><tr><td>ABHD16B</td><td>C20orf135</td><td>140701</td></tr><tr><td>ABHD17A</td><td>FAM108A1</td><td>81926</td></tr><tr><td>ABHD17B</td><td>FAM108B1</td><td>51104</td></tr><tr><td>ABHD17C</td><td>FAM108C1</td><td>58489</td></tr><tr><td>ABHD18</td><td>C4orf29</td><td>80167</td></tr><tr><td>ABRACL</td><td>C6orf115</td><td>58527</td></tr><tr><td>ACKR1</td><td>DARC</td><td>2532</td></tr><tr><td>ACKR2</td><td>CCBP2</td><td>1238</td></tr><tr><td>ACKR3</td><td>CXCR7</td><td>57007</td></tr><tr><td>ACKR4</td><td>CCRL1</td><td>51554</td></tr><tr><td>ACSM6</td><td>C10orf129</td><td>142827</td></tr><tr><td>ACTG1P4</td><td>LOC648740</td><td>648740</td></tr><tr><td>ACTL10</td><td>C20orf134</td><td>170487</td></tr></table></div>\n",
       "    <br />(rows: 2013, time: 6.1s,     4GB processed, job: job_Qogec9zEdrWGOVbaaDX_G3EioLM)<br />\n",
       "    <script>\n",
       "      require(['extensions/charting', 'element!3_146983056775', 'style!/static/extensions/charting.css'],\n",
       "        function(charts, dom) {\n",
       "          charts.render(dom,\n",
       "            {\n",
       "              chartStyle:\"paged_table\",\n",
       "              dataName:\"1\",\n",
       "              fields:\"HGNC_gene_symbol,original_gene_symbol,gene_id\",\n",
       "              totalRows:2013,\n",
       "              rowsPerPage:25,\n",
       "            }, {}, {\"rows\": [{\"c\": [{\"v\": \"A1BG-AS1\"}, {\"v\": \"NCRNA00181\"}, {\"v\": 503538}]}, {\"c\": [{\"v\": \"A2M-AS1\"}, {\"v\": \"LOC144571\"}, {\"v\": 144571}]}, {\"c\": [{\"v\": \"AACSP1\"}, {\"v\": \"AACSL\"}, {\"v\": 729522}]}, {\"c\": [{\"v\": \"AADACP1\"}, {\"v\": \"LOC201651\"}, {\"v\": 201651}]}, {\"c\": [{\"v\": \"AAED1\"}, {\"v\": \"C9orf21\"}, {\"v\": 195827}]}, {\"c\": [{\"v\": \"AAMDC\"}, {\"v\": \"C11orf67\"}, {\"v\": 28971}]}, {\"c\": [{\"v\": \"AAR2\"}, {\"v\": \"C20orf4\"}, {\"v\": 25980}]}, {\"c\": [{\"v\": \"AARD\"}, {\"v\": \"C8orf85\"}, {\"v\": 441376}]}, {\"c\": [{\"v\": \"AATBC\"}, {\"v\": \"LOC284837\"}, {\"v\": 284837}]}, {\"c\": [{\"v\": \"AATK-AS1\"}, {\"v\": \"LOC388428\"}, {\"v\": 388428}]}, {\"c\": [{\"v\": \"ABHD11-AS1\"}, {\"v\": \"WBSCR26\"}, {\"v\": 171022}]}, {\"c\": [{\"v\": \"ABHD16A\"}, {\"v\": \"BAT5\"}, {\"v\": 7920}]}, {\"c\": [{\"v\": \"ABHD16B\"}, {\"v\": \"C20orf135\"}, {\"v\": 140701}]}, {\"c\": [{\"v\": \"ABHD17A\"}, {\"v\": \"FAM108A1\"}, {\"v\": 81926}]}, {\"c\": [{\"v\": \"ABHD17B\"}, {\"v\": \"FAM108B1\"}, {\"v\": 51104}]}, {\"c\": [{\"v\": \"ABHD17C\"}, {\"v\": \"FAM108C1\"}, {\"v\": 58489}]}, {\"c\": [{\"v\": \"ABHD18\"}, {\"v\": \"C4orf29\"}, {\"v\": 80167}]}, {\"c\": [{\"v\": \"ABRACL\"}, {\"v\": \"C6orf115\"}, {\"v\": 58527}]}, {\"c\": [{\"v\": \"ACKR1\"}, {\"v\": \"DARC\"}, {\"v\": 2532}]}, {\"c\": [{\"v\": \"ACKR2\"}, {\"v\": \"CCBP2\"}, {\"v\": 1238}]}, {\"c\": [{\"v\": \"ACKR3\"}, {\"v\": \"CXCR7\"}, {\"v\": 57007}]}, {\"c\": [{\"v\": \"ACKR4\"}, {\"v\": \"CCRL1\"}, {\"v\": 51554}]}, {\"c\": [{\"v\": \"ACSM6\"}, {\"v\": \"C10orf129\"}, {\"v\": 142827}]}, {\"c\": [{\"v\": \"ACTG1P4\"}, {\"v\": \"LOC648740\"}, {\"v\": 648740}]}, {\"c\": [{\"v\": \"ACTL10\"}, {\"v\": \"C20orf134\"}, {\"v\": 170487}]}], \"cols\": [{\"type\": \"string\", \"id\": \"HGNC_gene_symbol\", \"label\": \"HGNC_gene_symbol\"}, {\"type\": \"string\", \"id\": \"original_gene_symbol\", \"label\": \"original_gene_symbol\"}, {\"type\": \"number\", \"id\": \"gene_id\", \"label\": \"gene_id\"}]});\n",
       "        }\n",
       "      );\n",
       "    </script>\n",
       "  "
      ],
      "text/plain": [
       "QueryResultsTable job_Qogec9zEdrWGOVbaaDX_G3EioLM"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "SELECT\n",
    "  HGNC_gene_symbol,\n",
    "  original_gene_symbol,\n",
    "  gene_id\n",
    "FROM\n",
    "  $mRNAseq_BQtable\n",
    "WHERE\n",
    "  ( original_gene_symbol IS NOT NULL\n",
    "    AND HGNC_gene_symbol IS NOT NULL\n",
    "    AND original_gene_symbol!=HGNC_gene_symbol\n",
    "    AND gene_id IS NOT NULL )\n",
    "GROUP BY\n",
    "  original_gene_symbol,\n",
    "  HGNC_gene_symbol,\n",
    "  gene_id\n",
    "ORDER BY\n",
    "  HGNC_gene_symbol"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "BigQuery is not just a \"look-up\" service -- you can also use it to perform calculations.  In this next query, we take a look at the mean, standard deviation, and coefficient of variation for the expression of EGFR, within each tumor-type, as well as the number of primary tumor samples that went into each summary statistic."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "    <div class=\"bqtv\" id=\"4_146983057033\"><table><tr><th>Study</th><th>n</th><th>exp_mean</th><th>exp_sigma</th><th>exp_cv</th></tr><tr><td>GBM</td><td>156</td><td>11.8977935121</td><td>2.37898710407</td><td>0.199951957617</td></tr><tr><td>SKCM</td><td>104</td><td>5.41189297221</td><td>2.3134143602</td><td>0.427468608873</td></tr><tr><td>BRCA</td><td>1095</td><td>7.3316523775</td><td>2.0186060124</td><td>0.275327567165</td></tr><tr><td>PCPG</td><td>179</td><td>5.21363387261</td><td>1.93511911652</td><td>0.371165134301</td></tr><tr><td>BLCA</td><td>408</td><td>9.26258869139</td><td>1.84003491022</td><td>0.19865233916</td></tr><tr><td>CESC</td><td>304</td><td>9.615363578</td><td>1.74670970515</td><td>0.181658206783</td></tr><tr><td>THYM</td><td>120</td><td>8.61408054406</td><td>1.74340125417</td><td>0.202389708948</td></tr><tr><td>LGG</td><td>516</td><td>11.1471994947</td><td>1.72222308803</td><td>0.154498274553</td></tr><tr><td>ESCA</td><td>184</td><td>11.6700063461</td><td>1.63953898439</td><td>0.140491696042</td></tr><tr><td>SARC</td><td>259</td><td>8.58641554296</td><td>1.59104767886</td><td>0.185298238933</td></tr><tr><td>UCEC</td><td>176</td><td>7.44476326481</td><td>1.5896445786</td><td>0.213525201818</td></tr><tr><td>LUSC</td><td>502</td><td>10.9294654726</td><td>1.54971809843</td><td>0.141792670677</td></tr><tr><td>TGCT</td><td>150</td><td>6.98336711449</td><td>1.50763881878</td><td>0.215889955958</td></tr><tr><td>KICH</td><td>66</td><td>9.48002047141</td><td>1.49846283983</td><td>0.158065359073</td></tr><tr><td>LUAD</td><td>515</td><td>9.89511374203</td><td>1.49825691249</td><td>0.151413814086</td></tr><tr><td>HNSC</td><td>520</td><td>11.3907589542</td><td>1.48880096929</td><td>0.130702526081</td></tr><tr><td>DLBC</td><td>48</td><td>4.41936166206</td><td>1.47405528888</td><td>0.333544842355</td></tr><tr><td>CHOL</td><td>36</td><td>9.19645454465</td><td>1.46185657708</td><td>0.158958712836</td></tr><tr><td>UCS</td><td>57</td><td>8.09853127117</td><td>1.44858302758</td><td>0.178869844306</td></tr><tr><td>ACC</td><td>79</td><td>7.7442559231</td><td>1.44367693285</td><td>0.186419063004</td></tr><tr><td>MESO</td><td>87</td><td>9.55197752348</td><td>1.43369934675</td><td>0.1500945059</td></tr><tr><td>STAD</td><td>415</td><td>10.2198608533</td><td>1.39368443653</td><td>0.136370196869</td></tr><tr><td>LIHC</td><td>371</td><td>9.60580817296</td><td>1.3719289882</td><td>0.142822859201</td></tr><tr><td>OV</td><td>305</td><td>8.53829302419</td><td>1.19815155796</td><td>0.140326825814</td></tr><tr><td>KIRP</td><td>290</td><td>9.36140534142</td><td>1.19513023402</td><td>0.127665685913</td></tr></table></div>\n",
       "    <br />(rows: 32, time: 2.3s,    11GB processed, job: job_uEvr7cSfInKOsx2ykx_7KH24Uuo)<br />\n",
       "    <script>\n",
       "      require(['extensions/charting', 'element!4_146983057033', 'style!/static/extensions/charting.css'],\n",
       "        function(charts, dom) {\n",
       "          charts.render(dom,\n",
       "            {\n",
       "              chartStyle:\"paged_table\",\n",
       "              dataName:\"2\",\n",
       "              fields:\"Study,n,exp_mean,exp_sigma,exp_cv\",\n",
       "              totalRows:32,\n",
       "              rowsPerPage:25,\n",
       "            }, {}, {\"rows\": [{\"c\": [{\"v\": \"GBM\"}, {\"v\": 156}, {\"v\": 11.897793512106745}, {\"v\": 2.3789871040733575}, {\"v\": 0.1999519576174011}]}, {\"c\": [{\"v\": \"SKCM\"}, {\"v\": 104}, {\"v\": 5.411892972210273}, {\"v\": 2.313414360197974}, {\"v\": 0.42746860887257193}]}, {\"c\": [{\"v\": \"BRCA\"}, {\"v\": 1095}, {\"v\": 7.3316523774978295}, {\"v\": 2.018606012396488}, {\"v\": 0.27532756716507123}]}, {\"c\": [{\"v\": \"PCPG\"}, {\"v\": 179}, {\"v\": 5.213633872613226}, {\"v\": 1.9351191165225574}, {\"v\": 0.3711651343005832}]}, {\"c\": [{\"v\": \"BLCA\"}, {\"v\": 408}, {\"v\": 9.262588691390237}, {\"v\": 1.840034910222928}, {\"v\": 0.19865233916013972}]}, {\"c\": [{\"v\": \"CESC\"}, {\"v\": 304}, {\"v\": 9.615363578003244}, {\"v\": 1.746709705149221}, {\"v\": 0.18165820678326844}]}, {\"c\": [{\"v\": \"THYM\"}, {\"v\": 120}, {\"v\": 8.614080544060982}, {\"v\": 1.74340125416887}, {\"v\": 0.2023897089482018}]}, {\"c\": [{\"v\": \"LGG\"}, {\"v\": 516}, {\"v\": 11.147199494698732}, {\"v\": 1.7222230880330807}, {\"v\": 0.1544982745533636}]}, {\"c\": [{\"v\": \"ESCA\"}, {\"v\": 184}, {\"v\": 11.670006346141202}, {\"v\": 1.6395389843853958}, {\"v\": 0.1404916960415814}]}, {\"c\": [{\"v\": \"SARC\"}, {\"v\": 259}, {\"v\": 8.586415542963953}, {\"v\": 1.5910476788622825}, {\"v\": 0.18529823893348016}]}, {\"c\": [{\"v\": \"UCEC\"}, {\"v\": 176}, {\"v\": 7.444763264806126}, {\"v\": 1.5896445786037432}, {\"v\": 0.21352520181783646}]}, {\"c\": [{\"v\": \"LUSC\"}, {\"v\": 502}, {\"v\": 10.929465472635274}, {\"v\": 1.5497180984335963}, {\"v\": 0.14179267067668716}]}, {\"c\": [{\"v\": \"TGCT\"}, {\"v\": 150}, {\"v\": 6.9833671144868354}, {\"v\": 1.5076388187833138}, {\"v\": 0.21588995595774302}]}, {\"c\": [{\"v\": \"KICH\"}, {\"v\": 66}, {\"v\": 9.480020471413471}, {\"v\": 1.4984628398310804}, {\"v\": 0.15806535907275943}]}, {\"c\": [{\"v\": \"LUAD\"}, {\"v\": 515}, {\"v\": 9.895113742029194}, {\"v\": 1.4982569124906273}, {\"v\": 0.15141381408551444}]}, {\"c\": [{\"v\": \"HNSC\"}, {\"v\": 520}, {\"v\": 11.390758954167762}, {\"v\": 1.488800969292324}, {\"v\": 0.13070252608124827}]}, {\"c\": [{\"v\": \"DLBC\"}, {\"v\": 48}, {\"v\": 4.419361662059624}, {\"v\": 1.4740552888799459}, {\"v\": 0.33354484235466914}]}, {\"c\": [{\"v\": \"CHOL\"}, {\"v\": 36}, {\"v\": 9.196454544646919}, {\"v\": 1.4618565770759486}, {\"v\": 0.15895871283644494}]}, {\"c\": [{\"v\": \"UCS\"}, {\"v\": 57}, {\"v\": 8.09853127117455}, {\"v\": 1.4485830275811267}, {\"v\": 0.17886984430585953}]}, {\"c\": [{\"v\": \"ACC\"}, {\"v\": 79}, {\"v\": 7.74425592310017}, {\"v\": 1.4436769328459804}, {\"v\": 0.18641906300380237}]}, {\"c\": [{\"v\": \"MESO\"}, {\"v\": 87}, {\"v\": 9.551977523482007}, {\"v\": 1.4336993467530101}, {\"v\": 0.15009450589979823}]}, {\"c\": [{\"v\": \"STAD\"}, {\"v\": 415}, {\"v\": 10.219860853268383}, {\"v\": 1.3936844365333605}, {\"v\": 0.13637019686893784}]}, {\"c\": [{\"v\": \"LIHC\"}, {\"v\": 371}, {\"v\": 9.605808172964679}, {\"v\": 1.3719289882009111}, {\"v\": 0.1428228592011834}]}, {\"c\": [{\"v\": \"OV\"}, {\"v\": 305}, {\"v\": 8.538293024186517}, {\"v\": 1.1981515579555688}, {\"v\": 0.14032682581419398}]}, {\"c\": [{\"v\": \"KIRP\"}, {\"v\": 290}, {\"v\": 9.361405341420824}, {\"v\": 1.1951302340237417}, {\"v\": 0.12766568591317415}]}], \"cols\": [{\"type\": \"string\", \"id\": \"Study\", \"label\": \"Study\"}, {\"type\": \"number\", \"id\": \"n\", \"label\": \"n\"}, {\"type\": \"number\", \"id\": \"exp_mean\", \"label\": \"exp_mean\"}, {\"type\": \"number\", \"id\": \"exp_sigma\", \"label\": \"exp_sigma\"}, {\"type\": \"number\", \"id\": \"exp_cv\", \"label\": \"exp_cv\"}]});\n",
       "        }\n",
       "      );\n",
       "    </script>\n",
       "  "
      ],
      "text/plain": [
       "QueryResultsTable job_uEvr7cSfInKOsx2ykx_7KH24Uuo"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%sql\n",
    "\n",
    "SELECT\n",
    "  Study,\n",
    "  n,\n",
    "  exp_mean,\n",
    "  exp_sigma,\n",
    "  (exp_sigma/exp_mean) AS exp_cv\n",
    "FROM (\n",
    "  SELECT\n",
    "    Study,\n",
    "    AVG(LOG2(normalized_count+1)) AS exp_mean,\n",
    "    STDDEV_POP(LOG2(normalized_count+1)) AS exp_sigma,\n",
    "    COUNT(AliquotBarcode) AS n\n",
    "  FROM\n",
    "    $mRNAseq_BQtable\n",
    "  WHERE\n",
    "    ( SampleTypeLetterCode=\"TP\"\n",
    "      AND HGNC_gene_symbol=\"EGFR\" )\n",
    "  GROUP BY\n",
    "    Study )\n",
    "ORDER BY\n",
    "  exp_sigma DESC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can also easily move the gene-symbol out of the WHERE clause and into the SELECT and GROUP BY clauses and have BigQuery do this same calculation over *all* genes and all tumor types.  This time we will use the `--module` option to define the query and then call it in the next cell from python."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%sql --module highVar\n",
    "\n",
    "SELECT\n",
    "  Study,\n",
    "  HGNC_gene_symbol,\n",
    "  n,\n",
    "  exp_mean,\n",
    "  exp_sigma,\n",
    "  (exp_sigma/exp_mean) AS exp_cv\n",
    "FROM (\n",
    "  SELECT\n",
    "    Study,\n",
    "    HGNC_gene_symbol,\n",
    "    AVG(LOG2(normalized_count+1)) AS exp_mean,\n",
    "    STDDEV_POP(LOG2(normalized_count+1)) AS exp_sigma,\n",
    "    COUNT(AliquotBarcode) AS n\n",
    "  FROM\n",
    "    $t\n",
    "  WHERE\n",
    "    ( SampleTypeLetterCode=\"TP\" )\n",
    "  GROUP BY\n",
    "    Study,\n",
    "    HGNC_gene_symbol )\n",
    "ORDER BY\n",
    "  exp_sigma DESC"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once we have defined a query, we can put it into a python object and print out the SQL statement to make sure it looks as expected:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SELECT\n",
      "  Study,\n",
      "  HGNC_gene_symbol,\n",
      "  n,\n",
      "  exp_mean,\n",
      "  exp_sigma,\n",
      "  (exp_sigma/exp_mean) AS exp_cv\n",
      "FROM (\n",
      "  SELECT\n",
      "    Study,\n",
      "    HGNC_gene_symbol,\n",
      "    AVG(LOG2(normalized_count+1)) AS exp_mean,\n",
      "    STDDEV_POP(LOG2(normalized_count+1)) AS exp_sigma,\n",
      "    COUNT(AliquotBarcode) AS n\n",
      "  FROM\n",
      "    [isb-cgc:tcga_201607_beta.mRNA_UNC_HiSeq_RSEM]\n",
      "  WHERE\n",
      "    ( SampleTypeLetterCode=\"TP\" )\n",
      "  GROUP BY\n",
      "    Study,\n",
      "    HGNC_gene_symbol )\n",
      "ORDER BY\n",
      "  exp_sigma DESC\n"
     ]
    }
   ],
   "source": [
    "q = bq.Query(highVar,t=mRNAseq_BQtable)\n",
    "print q.sql"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And then we can run it and save the results in another python object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "r = bq.Query(highVar,t=mRNAseq_BQtable).results()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#r.to_dataframe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since the result of the previous query is quite large (over 600,000 rows representing ~20,000 genes x ~30 tumor types), we might want to put those results into one or more subsequent queries that further refine these results, for example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%sql --module hv_genes\n",
    "\n",
    "SELECT *\n",
    "FROM ( $hv_result )\n",
    "HAVING\n",
    "  ( exp_mean > 6.\n",
    "    AND n >= 200\n",
    "    AND exp_cv > 0.5 )\n",
    "ORDER BY\n",
    "  exp_cv DESC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>HGNC_gene_symbol</th>\n",
       "      <th>n</th>\n",
       "      <th>exp_mean</th>\n",
       "      <th>exp_sigma</th>\n",
       "      <th>exp_cv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SARC</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>259</td>\n",
       "      <td>6.061873</td>\n",
       "      <td>5.548601</td>\n",
       "      <td>0.915328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>COAD</td>\n",
       "      <td>XIST</td>\n",
       "      <td>285</td>\n",
       "      <td>6.248359</td>\n",
       "      <td>5.470363</td>\n",
       "      <td>0.875488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>COAD</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>285</td>\n",
       "      <td>6.121045</td>\n",
       "      <td>5.155333</td>\n",
       "      <td>0.842231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>GSTM1</td>\n",
       "      <td>408</td>\n",
       "      <td>6.010392</td>\n",
       "      <td>5.006242</td>\n",
       "      <td>0.832931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KIRC</td>\n",
       "      <td>XIST</td>\n",
       "      <td>533</td>\n",
       "      <td>6.407005</td>\n",
       "      <td>5.271564</td>\n",
       "      <td>0.822781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>CPB1</td>\n",
       "      <td>1095</td>\n",
       "      <td>6.559260</td>\n",
       "      <td>5.168064</td>\n",
       "      <td>0.787904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>LGG</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>516</td>\n",
       "      <td>6.704050</td>\n",
       "      <td>5.271334</td>\n",
       "      <td>0.786291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>LGG</td>\n",
       "      <td>XIST</td>\n",
       "      <td>516</td>\n",
       "      <td>7.173960</td>\n",
       "      <td>5.539642</td>\n",
       "      <td>0.772187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HNSC</td>\n",
       "      <td>ACTC1</td>\n",
       "      <td>520</td>\n",
       "      <td>6.066815</td>\n",
       "      <td>4.653893</td>\n",
       "      <td>0.767106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>KRT6C</td>\n",
       "      <td>408</td>\n",
       "      <td>6.000740</td>\n",
       "      <td>4.567136</td>\n",
       "      <td>0.761095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LIHC</td>\n",
       "      <td>GSTM1</td>\n",
       "      <td>371</td>\n",
       "      <td>6.074415</td>\n",
       "      <td>4.605980</td>\n",
       "      <td>0.758259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SARC</td>\n",
       "      <td>PLA2G2A</td>\n",
       "      <td>259</td>\n",
       "      <td>6.002323</td>\n",
       "      <td>4.550597</td>\n",
       "      <td>0.758139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>STAD</td>\n",
       "      <td>REG3A</td>\n",
       "      <td>415</td>\n",
       "      <td>6.314804</td>\n",
       "      <td>4.661015</td>\n",
       "      <td>0.738109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>LUSC</td>\n",
       "      <td>MAGEA6</td>\n",
       "      <td>502</td>\n",
       "      <td>6.056004</td>\n",
       "      <td>4.452028</td>\n",
       "      <td>0.735143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SARC</td>\n",
       "      <td>XIST</td>\n",
       "      <td>259</td>\n",
       "      <td>7.028725</td>\n",
       "      <td>5.046253</td>\n",
       "      <td>0.717947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>STAD</td>\n",
       "      <td>FABP1</td>\n",
       "      <td>415</td>\n",
       "      <td>6.015140</td>\n",
       "      <td>4.304719</td>\n",
       "      <td>0.715647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>KIRC</td>\n",
       "      <td>FGB</td>\n",
       "      <td>533</td>\n",
       "      <td>6.385475</td>\n",
       "      <td>4.557211</td>\n",
       "      <td>0.713684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HNSC</td>\n",
       "      <td>ACTA1</td>\n",
       "      <td>520</td>\n",
       "      <td>6.272165</td>\n",
       "      <td>4.441155</td>\n",
       "      <td>0.708074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>KRT6B</td>\n",
       "      <td>408</td>\n",
       "      <td>6.821403</td>\n",
       "      <td>4.818030</td>\n",
       "      <td>0.706311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>STAD</td>\n",
       "      <td>RPS4Y1</td>\n",
       "      <td>415</td>\n",
       "      <td>6.695497</td>\n",
       "      <td>4.713168</td>\n",
       "      <td>0.703931</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CESC</td>\n",
       "      <td>SPRR2E</td>\n",
       "      <td>304</td>\n",
       "      <td>6.072856</td>\n",
       "      <td>4.266463</td>\n",
       "      <td>0.702546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>LIHC</td>\n",
       "      <td>FXYD2</td>\n",
       "      <td>371</td>\n",
       "      <td>6.026990</td>\n",
       "      <td>4.227431</td>\n",
       "      <td>0.701417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>CLCA4</td>\n",
       "      <td>408</td>\n",
       "      <td>6.096983</td>\n",
       "      <td>4.248395</td>\n",
       "      <td>0.696803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>LIHC</td>\n",
       "      <td>EPCAM</td>\n",
       "      <td>371</td>\n",
       "      <td>6.018766</td>\n",
       "      <td>4.176006</td>\n",
       "      <td>0.693831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>KRT4</td>\n",
       "      <td>408</td>\n",
       "      <td>6.182480</td>\n",
       "      <td>4.287769</td>\n",
       "      <td>0.693536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>LIHC</td>\n",
       "      <td>CYP1A2</td>\n",
       "      <td>371</td>\n",
       "      <td>6.324730</td>\n",
       "      <td>4.386093</td>\n",
       "      <td>0.693483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>LUSC</td>\n",
       "      <td>MAGEA9B</td>\n",
       "      <td>502</td>\n",
       "      <td>6.100760</td>\n",
       "      <td>4.227037</td>\n",
       "      <td>0.692871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>LUAD</td>\n",
       "      <td>XIST</td>\n",
       "      <td>515</td>\n",
       "      <td>7.728633</td>\n",
       "      <td>5.342959</td>\n",
       "      <td>0.691320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>KIRC</td>\n",
       "      <td>KDM5D</td>\n",
       "      <td>533</td>\n",
       "      <td>6.093867</td>\n",
       "      <td>4.209498</td>\n",
       "      <td>0.690776</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SARC</td>\n",
       "      <td>CHRDL2</td>\n",
       "      <td>259</td>\n",
       "      <td>6.360666</td>\n",
       "      <td>4.375406</td>\n",
       "      <td>0.687885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>STAD</td>\n",
       "      <td>TM4SF4</td>\n",
       "      <td>415</td>\n",
       "      <td>6.620646</td>\n",
       "      <td>3.355416</td>\n",
       "      <td>0.506811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>LIHC</td>\n",
       "      <td>HGFAC</td>\n",
       "      <td>371</td>\n",
       "      <td>8.064111</td>\n",
       "      <td>4.084009</td>\n",
       "      <td>0.506443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>STAD</td>\n",
       "      <td>CDX1</td>\n",
       "      <td>415</td>\n",
       "      <td>7.238605</td>\n",
       "      <td>3.665532</td>\n",
       "      <td>0.506387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>KIRP</td>\n",
       "      <td>HBA1</td>\n",
       "      <td>290</td>\n",
       "      <td>6.039078</td>\n",
       "      <td>3.055942</td>\n",
       "      <td>0.506028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>PRAD</td>\n",
       "      <td>CST1</td>\n",
       "      <td>497</td>\n",
       "      <td>6.162322</td>\n",
       "      <td>3.117676</td>\n",
       "      <td>0.505926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>OV</td>\n",
       "      <td>GAL3ST3</td>\n",
       "      <td>305</td>\n",
       "      <td>6.111476</td>\n",
       "      <td>3.089889</td>\n",
       "      <td>0.505588</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>COAD</td>\n",
       "      <td>ALDH1L1</td>\n",
       "      <td>285</td>\n",
       "      <td>6.285763</td>\n",
       "      <td>3.170998</td>\n",
       "      <td>0.504473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>KLHDC7A</td>\n",
       "      <td>1095</td>\n",
       "      <td>6.339753</td>\n",
       "      <td>3.197413</td>\n",
       "      <td>0.504343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>LUSC</td>\n",
       "      <td>SBSN</td>\n",
       "      <td>502</td>\n",
       "      <td>6.835398</td>\n",
       "      <td>3.444914</td>\n",
       "      <td>0.503982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>KRT81</td>\n",
       "      <td>1095</td>\n",
       "      <td>6.511374</td>\n",
       "      <td>3.280551</td>\n",
       "      <td>0.503819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>SARC</td>\n",
       "      <td>TPSAB1</td>\n",
       "      <td>259</td>\n",
       "      <td>6.814440</td>\n",
       "      <td>3.433117</td>\n",
       "      <td>0.503800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>KIRP</td>\n",
       "      <td>ALDOB</td>\n",
       "      <td>290</td>\n",
       "      <td>6.118838</td>\n",
       "      <td>3.080089</td>\n",
       "      <td>0.503378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>SARC</td>\n",
       "      <td>SERPINA3</td>\n",
       "      <td>259</td>\n",
       "      <td>6.266452</td>\n",
       "      <td>3.154244</td>\n",
       "      <td>0.503354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>BRCA</td>\n",
       "      <td>ELF5</td>\n",
       "      <td>1095</td>\n",
       "      <td>6.514652</td>\n",
       "      <td>3.279154</td>\n",
       "      <td>0.503351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>LIHC</td>\n",
       "      <td>SCGN</td>\n",
       "      <td>371</td>\n",
       "      <td>6.090981</td>\n",
       "      <td>3.063063</td>\n",
       "      <td>0.502885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>CYP4F22</td>\n",
       "      <td>408</td>\n",
       "      <td>6.981163</td>\n",
       "      <td>3.508778</td>\n",
       "      <td>0.502606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>PRAD</td>\n",
       "      <td>ORM2</td>\n",
       "      <td>497</td>\n",
       "      <td>6.222926</td>\n",
       "      <td>3.123566</td>\n",
       "      <td>0.501945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>CESC</td>\n",
       "      <td>MMP10</td>\n",
       "      <td>304</td>\n",
       "      <td>6.789319</td>\n",
       "      <td>3.405935</td>\n",
       "      <td>0.501661</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>LUAD</td>\n",
       "      <td>B3GNT6</td>\n",
       "      <td>515</td>\n",
       "      <td>6.068778</td>\n",
       "      <td>3.042877</td>\n",
       "      <td>0.501399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>STAD</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>415</td>\n",
       "      <td>6.206014</td>\n",
       "      <td>3.110105</td>\n",
       "      <td>0.501144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>SARC</td>\n",
       "      <td>SMOC1</td>\n",
       "      <td>259</td>\n",
       "      <td>7.165206</td>\n",
       "      <td>3.589628</td>\n",
       "      <td>0.500980</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>SARC</td>\n",
       "      <td>HR</td>\n",
       "      <td>259</td>\n",
       "      <td>6.015107</td>\n",
       "      <td>3.013293</td>\n",
       "      <td>0.500954</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>SARC</td>\n",
       "      <td>ABCA8</td>\n",
       "      <td>259</td>\n",
       "      <td>6.951896</td>\n",
       "      <td>3.482219</td>\n",
       "      <td>0.500902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>CESC</td>\n",
       "      <td>SPRR1B</td>\n",
       "      <td>304</td>\n",
       "      <td>9.031974</td>\n",
       "      <td>4.524126</td>\n",
       "      <td>0.500901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>320</th>\n",
       "      <td>KIRC</td>\n",
       "      <td>SLC5A8</td>\n",
       "      <td>533</td>\n",
       "      <td>7.190818</td>\n",
       "      <td>3.598825</td>\n",
       "      <td>0.500475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>321</th>\n",
       "      <td>OV</td>\n",
       "      <td>ZNHIT2</td>\n",
       "      <td>305</td>\n",
       "      <td>6.444236</td>\n",
       "      <td>3.224612</td>\n",
       "      <td>0.500387</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>DES</td>\n",
       "      <td>408</td>\n",
       "      <td>8.655334</td>\n",
       "      <td>4.330614</td>\n",
       "      <td>0.500341</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>323</th>\n",
       "      <td>LIHC</td>\n",
       "      <td>PLPP2</td>\n",
       "      <td>371</td>\n",
       "      <td>6.643719</td>\n",
       "      <td>3.323184</td>\n",
       "      <td>0.500199</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>324</th>\n",
       "      <td>THCA</td>\n",
       "      <td>DCSTAMP</td>\n",
       "      <td>505</td>\n",
       "      <td>8.456327</td>\n",
       "      <td>4.228831</td>\n",
       "      <td>0.500079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>325</th>\n",
       "      <td>BLCA</td>\n",
       "      <td>CEACAM6</td>\n",
       "      <td>408</td>\n",
       "      <td>7.919889</td>\n",
       "      <td>3.960153</td>\n",
       "      <td>0.500026</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>326 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Study HGNC_gene_symbol     n  exp_mean  exp_sigma    exp_cv\n",
       "0    SARC           RPS4Y1   259  6.061873   5.548601  0.915328\n",
       "1    COAD             XIST   285  6.248359   5.470363  0.875488\n",
       "2    COAD           RPS4Y1   285  6.121045   5.155333  0.842231\n",
       "3    BLCA            GSTM1   408  6.010392   5.006242  0.832931\n",
       "4    KIRC             XIST   533  6.407005   5.271564  0.822781\n",
       "5    BRCA             CPB1  1095  6.559260   5.168064  0.787904\n",
       "6     LGG           RPS4Y1   516  6.704050   5.271334  0.786291\n",
       "7     LGG             XIST   516  7.173960   5.539642  0.772187\n",
       "8    HNSC            ACTC1   520  6.066815   4.653893  0.767106\n",
       "9    BLCA            KRT6C   408  6.000740   4.567136  0.761095\n",
       "10   LIHC            GSTM1   371  6.074415   4.605980  0.758259\n",
       "11   SARC          PLA2G2A   259  6.002323   4.550597  0.758139\n",
       "12   STAD            REG3A   415  6.314804   4.661015  0.738109\n",
       "13   LUSC           MAGEA6   502  6.056004   4.452028  0.735143\n",
       "14   SARC             XIST   259  7.028725   5.046253  0.717947\n",
       "15   STAD            FABP1   415  6.015140   4.304719  0.715647\n",
       "16   KIRC              FGB   533  6.385475   4.557211  0.713684\n",
       "17   HNSC            ACTA1   520  6.272165   4.441155  0.708074\n",
       "18   BLCA            KRT6B   408  6.821403   4.818030  0.706311\n",
       "19   STAD           RPS4Y1   415  6.695497   4.713168  0.703931\n",
       "20   CESC           SPRR2E   304  6.072856   4.266463  0.702546\n",
       "21   LIHC            FXYD2   371  6.026990   4.227431  0.701417\n",
       "22   BLCA            CLCA4   408  6.096983   4.248395  0.696803\n",
       "23   LIHC            EPCAM   371  6.018766   4.176006  0.693831\n",
       "24   BLCA             KRT4   408  6.182480   4.287769  0.693536\n",
       "25   LIHC           CYP1A2   371  6.324730   4.386093  0.693483\n",
       "26   LUSC          MAGEA9B   502  6.100760   4.227037  0.692871\n",
       "27   LUAD             XIST   515  7.728633   5.342959  0.691320\n",
       "28   KIRC            KDM5D   533  6.093867   4.209498  0.690776\n",
       "29   SARC           CHRDL2   259  6.360666   4.375406  0.687885\n",
       "..    ...              ...   ...       ...        ...       ...\n",
       "296  STAD           TM4SF4   415  6.620646   3.355416  0.506811\n",
       "297  LIHC            HGFAC   371  8.064111   4.084009  0.506443\n",
       "298  STAD             CDX1   415  7.238605   3.665532  0.506387\n",
       "299  KIRP             HBA1   290  6.039078   3.055942  0.506028\n",
       "300  PRAD             CST1   497  6.162322   3.117676  0.505926\n",
       "301    OV          GAL3ST3   305  6.111476   3.089889  0.505588\n",
       "302  COAD          ALDH1L1   285  6.285763   3.170998  0.504473\n",
       "303  BRCA          KLHDC7A  1095  6.339753   3.197413  0.504343\n",
       "304  LUSC             SBSN   502  6.835398   3.444914  0.503982\n",
       "305  BRCA            KRT81  1095  6.511374   3.280551  0.503819\n",
       "306  SARC           TPSAB1   259  6.814440   3.433117  0.503800\n",
       "307  KIRP            ALDOB   290  6.118838   3.080089  0.503378\n",
       "308  SARC         SERPINA3   259  6.266452   3.154244  0.503354\n",
       "309  BRCA             ELF5  1095  6.514652   3.279154  0.503351\n",
       "310  LIHC             SCGN   371  6.090981   3.063063  0.502885\n",
       "311  BLCA          CYP4F22   408  6.981163   3.508778  0.502606\n",
       "312  PRAD             ORM2   497  6.222926   3.123566  0.501945\n",
       "313  CESC            MMP10   304  6.789319   3.405935  0.501661\n",
       "314  LUAD           B3GNT6   515  6.068778   3.042877  0.501399\n",
       "315  STAD           CYP2B6   415  6.206014   3.110105  0.501144\n",
       "316  SARC            SMOC1   259  7.165206   3.589628  0.500980\n",
       "317  SARC               HR   259  6.015107   3.013293  0.500954\n",
       "318  SARC            ABCA8   259  6.951896   3.482219  0.500902\n",
       "319  CESC           SPRR1B   304  9.031974   4.524126  0.500901\n",
       "320  KIRC           SLC5A8   533  7.190818   3.598825  0.500475\n",
       "321    OV           ZNHIT2   305  6.444236   3.224612  0.500387\n",
       "322  BLCA              DES   408  8.655334   4.330614  0.500341\n",
       "323  LIHC            PLPP2   371  6.643719   3.323184  0.500199\n",
       "324  THCA          DCSTAMP   505  8.456327   4.228831  0.500079\n",
       "325  BLCA          CEACAM6   408  7.919889   3.960153  0.500026\n",
       "\n",
       "[326 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bq.Query(hv_genes,hv_result=r).results().to_dataframe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
